platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/iss/DefaultIssHandler.java

```diff
@@ -0,0 +1,201 @@
+    private boolean halted;
+
+    /**
+     * Create an object responsible for handling ISS events.
+     *
+     * @param platformContext       the platform context
+     * @param haltRequestedConsumer consumer to invoke when a system halt is desired
+     * @param fatalErrorConsumer    consumer to invoke if a fatal error occurs
+     * @param issScratchpad         scratchpad for ISS data, is persistent across restarts
+     */
+    public DefaultIssHandler(
+            @NonNull final PlatformContext platformContext,
+            @NonNull final Consumer<String> haltRequestedConsumer,
+            @NonNull final FatalErrorConsumer fatalErrorConsumer,
+            @NonNull final Scratchpad<IssScratchpad> issScratchpad) {
+        this.haltRequestedConsumer = Objects.requireNonNull(haltRequestedConsumer);
+        this.fatalErrorConsumer = Objects.requireNonNull(fatalErrorConsumer);
+        this.stateConfig = platformContext.getConfiguration().getConfigData(StateConfig.class);
+        this.issScratchpad = Objects.requireNonNull(issScratchpad);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void issObserved(@NonNull final IssNotification issNotification) {
+        switch (issNotification.getIssType()) {
+            case SELF_ISS -> selfIssObserver(issNotification.getRound());
+            case OTHER_ISS -> otherIss();
+            case CATASTROPHIC_ISS -> catastrophicIssObserver(issNotification.getRound());
+        }
+    }
+
+    /**
+     * This method is called whenever any node is observed in disagreement with the consensus hash.
+     */
+    private void otherIss() {
+        if (halted) {
+            // don't take any action once halted
+            return;
+        }
+        if (stateConfig.haltOnAnyIss()) {
+            haltRequestedConsumer.accept("other node observed with ISS");
+            halted = true;
+        }
+    }
+
+    /**
+     * Record the latest ISS round in the scratchpad. Does nothing if this is not the latest ISS that has been
+     * observed.
+     *
+     * @param issRound the round of the observed ISS
+     */
+    private void updateIssRoundInScratchpad(final long issRound) {
+        issScratchpad.atomicOperation(data -> {
+            final SerializableLong lastIssRound = (SerializableLong) data.get(IssScratchpad.LAST_ISS_ROUND);
+
+            if (lastIssRound == null || lastIssRound.getValue() < issRound) {
+                data.put(IssScratchpad.LAST_ISS_ROUND, new SerializableLong(issRound));
+                return true;
+            }
+
+            // Data was not modified, no need to flush data to disk. It is possible to observe ISS rounds out of order,
+            // and we only want to increment this number.
+            return false;
+        });
+    }
+
+    /**
+     * This method is called when there is a self ISS.
+     *
+     * @param round the round of the ISS
+     */
+    private void selfIssObserver(@NonNull final Long round) {
+
+        if (halted) {
```

Comment:
Can we ignore concurrency here or should halted be refactored to be atomic?

Reply:
Concurrency is intentionally ignored at this scope.

The ISS handler is a direct scheduler (meaning it is executed on the caller's thread), and the caller is a sequential scheduler (meaning there is a single logical thread). 

The goal of the wiring framework is to make business logic like this entirely devoid of threading primitives, and to handle all threading concerns at the wiring framework level. In current platform business logic, if you see things like atomic variables, locks, or the `synchronized` keyword, it usually means it is legacy code that we haven't gotten around to refactoring.